history api

canonici opened the pull request:
requires https://github.com/geneweb/geneweb/pull/1803

Commit:
history api

ilankri commented on src/assets/api.proto:
```diff
@@ -759,3 +759,59 @@ message Error {
+    repeated HistoryEntry entries = 1;
+    required int32 page = 2;
+    required int32 page_max = 3;
+}
```
Could you please add a newline at the end of the file?

ilankri commented on src/api.ml:
```diff
@@ -707,3 +707,126 @@ let print_all_families conf base =
+      }
+    | None -> None
+
+  let history_entry _conf base time user action keyo =
```
```suggestion
  let history_entry base time user action keyo =
```

ilankri commented on src/api.ml:
```diff
@@ -707,3 +707,126 @@ let print_all_families conf base =
+       day = Int32.of_int 0;
+       hour = Int32.of_int 0;
+       minute = Int32.of_int 0;
+       second = Int32.of_int 0}
```
```suggestion
      {M.Time.year = Int32.zero;
       month = Int32.zero;
       day = Int32.zero;
       hour = Int32.zero;
       minute = Int32.zero;
       second = Int32.zero}
```

ilankri commented on src/api.ml:
```diff
@@ -707,3 +707,126 @@ let print_all_families conf base =
+  let person_of_key base s =
+    let year_of_date = function
+      | Adef.Cdate (Dgreg (dmy, _cal)) -> Some (Int32.of_int dmy.year)
+      | _ -> None
```
Could you please enumerate all the cases?

ilankri commented on src/api.ml:
```diff
@@ -707,3 +707,126 @@ let print_all_families conf base =
+    in
+    let date_of_death = function
+      | Def.Death (_, cdate) -> cdate
+      | _ -> Adef.Cnone
```
Idem

ilankri commented on src/api.ml:
```diff
@@ -707,3 +707,126 @@ let print_all_families conf base =
+    | Some ip ->
+      let pers = Gwdb.poi base ip in
+      let firstname = Gwdb.sou base (Gwdb.get_surname pers) in
+      let lastname = Gwdb.sou base (Gwdb.get_first_name pers) in
```
```suggestion
      let lastname = Gwdb.sou base (Gwdb.get_surname pers) in
      let firstname = Gwdb.sou base (Gwdb.get_first_name pers) in
```

ilankri commented on src/api.ml:
```diff
@@ -707,3 +707,126 @@ let print_all_families conf base =
+        birth_year;
+        death_year;
+      }
+    | None -> None
```
Could you please use `Option.map`?

ilankri commented on src/api.ml:
```diff
@@ -707,3 +707,126 @@ let print_all_families conf base =
+  let history_entry _conf base time user action keyo =
+    let time = time_of_string time in
+    let action = action_of_string action in
+    let person = Option.join (Option.map (person_of_key base) keyo) in
```
```suggestion
    let person = Option.bind keyo (person_of_key base) in
```

ilankri commented on src/api.ml:
```diff
@@ -707,3 +707,126 @@ let print_all_families conf base =
+      |> List.map (fun (time, user, action, keyo) ->
+          history_entry conf base time user action keyo)
+    in
+    Mext.gen_history {M.History.entries; page; page_max;}
```
```suggestion
    Mext.gen_history {M.History.entries; page; page_max}
```

ilankri commented on src/api.ml:
```diff
@@ -707,3 +707,126 @@ let print_all_families conf base =
+        let rec aux res l n =
+          match l with
+          | x :: l when n > 0 -> aux (x :: res) l (n - 1)
+          | _ :: _ when n = 0 -> res
+          | _ -> res
```
```suggestion
          | _ -> res
```

ilankri commented on src/api.ml:
```diff
@@ -707,3 +707,126 @@ let print_all_families conf base =
+      let rec ntail l n = match l with
+        | _ :: l when n > 0 -> ntail l (n - 1)
+        | _ :: l when n = 0 -> l
+        | _ -> l
```
```suggestion
        | _ -> l
```

so that for all `l`, `ntail l 0` = `l` (assumption: the first page has number 1 and not 0, otherwise, we should not apply this change).

ilankri commented on src/api.ml:
```diff
@@ -707,3 +707,126 @@ let print_all_families conf base =
+        | _ :: l when n = 0 -> l
+        | _ -> l
+      in
+      let nhd l n =
```
```suggestion
      let take l n =
```

and this function should probably be moved to a separate module in the directory `lib/util/` of https://github.com/geneweb/geneweb.

ilankri commented on src/api.ml:
```diff
@@ -707,3 +707,126 @@ let print_all_families conf base =
+    in
+    let entries = Geneweb.History.map_history conf base f in
+    let sublist l pos len =
+      let rec ntail l n = match l with
```
```suggestion
      let rec drop l n = match l with
```

and this function should probably be moved to a separate module in the directory `lib/util/` of https://github.com/geneweb/geneweb.

ilankri committed:
Cleanup: Add a newline at the end of a file

ilankri committed:
Build: Make the code compile

Adapt the code to the following change:

https://github.com/geneweb/geneweb/pull/1803/commits/c02807512ad6f0dfbd9ff4b26f9c33f88cd0f168

ilankri committed:
Cleanup: Remove an unused function parameter

ilankri committed:
Cleanup: Replace `Int32.of_int 0` with `Int32.zero`

ilankri committed:
Cleanup: Make some pattern matching expressions more robust

ilankri committed:
History: Fix mixed-up identifiers

ilankri committed:
Cleanup: Use the standard library function `Option.map`

ilankri committed:
Cleanup: Remove a useless semicolon

ilankri committed:
History: Simplify some code

ilankri commented on src/api.ml:
```diff
@@ -707,3 +707,129 @@ let print_all_families conf base =
+      time, user, action, keyo
+    in
+    let entries = Geneweb.History.map_history conf f in
+    let sublist l pos len =
```
This function should probably be moved to a separate module in the directory `lib/util/` of https://github.com/geneweb/geneweb.

ilankri committed:
History: Use the library function `Ext_list.sublist` of `Geneweb`

Commit:
return split key when person does not exist in the base

Commit:
returns whether person exists in base

Commit:
returns whether the person has an history

Commit:
filter on user name

Commit:
reverse order of history entries

ilankri commented on src/api.ml:
```diff
@@ -707,3 +707,142 @@ let print_all_families conf base =
+    | "cs" -> `source_modified
+    | "co" -> `occupation_modified
+    | s -> raise (Invalid_argument s)
+  
```
Could you please remove the trailing white spaces?

ilankri commented on src/api.ml:
```diff
@@ -825,12 +825,15 @@ module HistoryApi = struct
+      |> List.filter_map (fun (time, user, action, keyo) ->
+          if not filter || user = Option.get filter_user then
+            Some (history_entry conf base time user action keyo)
+          else None)
```
Could you please rewrite this using the functions `Option.bind` and `Geneweb.Ext_option.return_if`?

ilankri commented on src/api.ml:
```diff
@@ -835,6 +835,7 @@ module HistoryApi = struct
             Some (history_entry conf base time user action keyo)
           else None)
     in
+    let entries = List.rev entries in
```
Don't we rather want to reverse the list of the entries before extracting the sublist?

Commit:
fix year_of_date

Commit:
allow friends to access history api

Commit:
history: return burial/baptism year when there is no birth/death

Commit:
history: return total number of elements

Commit:
only build history entry on filtered entry

Commit:
history: return note data

ilankri commented on src/api.ml:
```diff
@@ -804,40 +804,75 @@ module HistoryApi = struct
+      else link_parameters in
+    let link_txt =
+      if pg <> "" then
+        Printf.sprintf {|[%s]|} pg
```
Is there a particular reason for using the alternative syntax for this string literal?

ilankri commented on src/api.ml:
```diff
@@ -804,40 +804,75 @@ module HistoryApi = struct
+      else link_parameters in
+    let link_parameters =
+      if Option.is_some part then
+        Printf.sprintf {|%s;v=%d|} link_parameters (Option.get part)
```
Idem

ilankri commented on src/api.ml:
```diff
@@ -804,40 +804,75 @@ module HistoryApi = struct
+    let link_parameters = "m=NOTES" in
+    let link_parameters =
+      if pg <> "" then
+        Printf.sprintf {|%s;f=%s|} link_parameters pg
```
Idem

ilankri commented on src/api.ml:
```diff
@@ -707,3 +707,179 @@ let print_all_families conf base =
+        Printf.sprintf {|%s;f=%s|} link_parameters pg
+      else link_parameters in
+    let link_parameters =
+      if Option.is_some part then
+        Printf.sprintf {|%s;v=%d|} link_parameters (Option.get part)
+      else link_parameters in
```
```suggestion
      match part with
      | Some part -> Printf.sprintf {|%s;v=%d|} link_parameters part
      | None -> link_parameters in
```

ilankri replied:
@canonici?

ilankri commented on src/api.ml:
```diff
@@ -707,3 +707,179 @@ let print_all_families conf base =
+    in
+    let ipage = Int32.to_int page in
+    let elements_per_page = Int32.to_int elements_per_page in
+    let filter = if Option.is_some filter_user then
+        let filter_user = Option.get filter_user in
+        fun ~time:_ ~user ~action:_ ~keyo:_ -> filter_user = user
+      else fun ~time:_ ~user:_ ~action:_ ~keyo:_ -> true
```
```suggestion
    let filter = match filter_user with
        | Some filter_user ->
             fun ~time:_ ~user ~action:_ ~keyo:_ -> filter_user = user
        | None -> fun ~time:_ ~user:_ ~action:_ ~keyo:_ -> true
```
